### PR TITLE
add bearer token on signout 

### PR DESCRIPTION
**Pull request by yuenmichelle1**
Related: https://github.com/zooniverse/panoptes/pull/4134
https://github.com/zooniverse/how-to-zooniverse/issues/319

Edit: Add missing Bearer Authorization Header on Signout

### CODE REVIEW

**Review by eatyourgreens — APPROVED** [review]
LGTM. Thanks for this!

**Review by camallen — COMMENTED** [review]
Would it be better to ensure the credentialed superagent singleton has the access token setup all the time as well vs injecting it on this action? Longer term that means less changes to the client if we need this behaviour on other actions.

Similar to how we do that here? https://github.com/zooniverse/panoptes-javascript-client/blob/87f1d174953ccec7a9f86fbbb5dfd387372905b8/lib/auth.js#L51-L52

That api client is the same under the hood pretty much, e.g. https://github.com/zooniverse/panoptes-javascript-client/blob/87f1d174953ccec7a9f86fbbb5dfd387372905b8/lib/api-client.js#L4-L8

That said if we don't need it then this change looks fine to me but i'd like a review from someone else to confirm it's acceptable longer term.

**Comment by eatyourgreens** [comment]
`superagent.auth()` supports bearer tokens now, so that’s an option. 
https://ladjs.github.io/superagent/#authentication

The agents themselves aren’t exposed publicly by `json-api-client` so the headers would still have to be constructed manually before calling `makeCredentialHTTPRequest`. 

Does the `signOut` method in `oauth.js` need to be updated too?
https://github.com/zooniverse/panoptes-javascript-client/blob/87f1d174953ccec7a9f86fbbb5dfd387372905b8/lib/oauth.js#L129-L132

**Comment by yuenmichelle1** [comment]
Probably could set `.auth` somewhere around here:

**Comment by yuenmichelle1** [comment]
If we wanted to set `auth` on credentialed superagent then we could possibly do it here:
https://github.com/zooniverse/panoptes-javascript-client/blob/master/lib/json-api-client/make-http-request.js#L26-L32. I think? I havent checked. 

One thing I'm unsure of is when grabbing a new token: 
https://github.com/zooniverse/panoptes-javascript-client/blob/master/lib/auth.js#L34-L38 when a token is already set on superagent.auth request, will it set the old token or the new one.

**Comment by eatyourgreens** [comment]
>One thing I'm unsure of is when grabbing a new token:
https://github.com/zooniverse/panoptes-javascript-client/blob/master/lib/auth.js#L34-L38 when a token is already set on superagent.auth request, will it set the old token or the new one.

Ages since I've looked at this code, so I'm not sure, but I think it refreshes the bearer token if the existing token is within five minutes of expiring. If you've cleared the stored token, then it should always fetch a new one from Panoptes.

**Comment by eatyourgreens** [comment]
>That said if we don't need it then this change looks fine to me but i'd like a review from someone else to confirm it's acceptable longer term.

I already reviewed this, last week. Still looks good to me, so I'm happy for it to be merged as-is. `oauth.signOut()` might need the same changes to the request headers, so that this works for https://classroom.zooniverse.org and other domains that use the `oauth` client.

**Comment by yuenmichelle1** [comment]
> I already reviewed this, last week. Still looks good to me, so I'm happy for it to be merged as-is. oauth.signOut() might need the same changes to the request headers, so that this works for https://classroom.zooniverse.org/ and other domains that use the oauth client. 

Apologies @eatyourgreens , Mondays at Adler are tricky since meetings are stacked together. I wanted to test around with setting the header somewhere higher on chain. Will ping you again for another review if needed. 

Otherwise, I'll go with original fix (and update oauth client with a similar fix) and get this merged. Thanks!

**Comment by eatyourgreens** [comment]
No worries, just wanted to make sure I'm not holding this up.

File: lib/auth.js [hunk 1]
```diff
@@ -50,6 +50,7 @@ const authClient = new Model({
 
     this._bearerToken = response.access_token;
     apiClient.headers.Authorization = 'Bearer ' + this._bearerToken;
+    JSONAPIClient.headers.Authorization = 'Bearer ' + this._bearerToken;
```
Comment (eatyourgreens) [inline lib/auth.js:53]: `JSONAPICLient` is a class, not an instance of a client. New clients are instantiated with:
```js
const client = new JSONAPIClient(APIRootURL, headers, mixins)
```

You can pass default headers to the constructor, then access them from `client.headers` (I think.)

See the Panoptes API client for an example.
https://github.com/zooniverse/panoptes-javascript-client/blob/87f1d174953ccec7a9f86fbbb5dfd387372905b8/lib/api-client.js#L4-L48

The auth client doesn't instantiate an API client, so you'd have to rewrite it to do that before making this change.

Comment (yuenmichelle1) [reply lib/auth.js]: oh crud. oops you are right.

**Comment by yuenmichelle1** [comment]
Hey @eatyourgreens , do you mind re-reviewing? I ended up reverting the change on auth.js and did the set of token on deleteHeaders. 
Mainly, I wanted to double check on `oauth.js` (not sure where that is used, ), if I'm grabbing the token correctly.

**Comment by eatyourgreens** [comment]
>Mainly, I wanted to double check on oauth.js (not sure where that is used, ), if I'm grabbing the token correctly.

I'm not 100% sure how `oauth.js` stores the token either, but the changes look reasonable to me. https://classroom.zooniverse.org uses that client for Panoptes auth. It's also used by any project that's not hosted on a `*.zooniverse.org` domain eg. Scribes of the Cairo Genizah.

File: lib/oauth.js [hunk 2]
```diff
@@ -131,6 +131,10 @@ const authClient = new Model({
             ['X-CSRF-Token']: token
           };
 
+          if (this._tokenDetails && this._tokenDetails.access_token) {
+            deleteHeaders['Authorization'] = 'Bearer ' + this._tokenDetails.access_token
+          }
```
Comment (eatyourgreens) [inline lib/oauth.js:136]: Similar comment here. This code is absolutely fine but if the token details are missing, then you aren't logged in anway (I think.)

**Pull request merged**